Remove Sum Variance

JoostJM opened the pull request:
Remove Sum Variance (GLCM), as this feature is mathematically equal to Cluster Tendency. Document the proof of this equality in a new documentation section (Excluded Radiomic Features).

As this proof is coded in LaTeX, here is a render of the proof:
![proof_sumvar_clusten](https://user-images.githubusercontent.com/18026947/30038240-66072df0-9190-11e7-8c04-da0289718306.png)

Aside from this equality, there is also a near equality of Joint Average and Sum Average  (when GLCM is symmetrical, Sum Average = 2 * Joint Average). Therefore, disable Sum Average by default in the example parameter files and add a warning to the docstring of Sum Average.

Finally, there was a small error in the documentation of the GLCM formulas (u_x and u_y erroneously represented as a function of i and j, respectively (they are scalar values)).

cc @Radiomics/developers

clarehchao commented:
Hi all,

Thanks for making this nice repo for extracting radiomic features with such
ease and flexibility!

I was going through the documentation on how the GLCM-related features are
defined and found some inconsistency that I was wondering if someone can
help clarify since this pull request was regarding the feature definition.

Sum variance:

   - sum variance was defined slightly differently in Aert et al., 2014
   (Nature Communications, supplementary documentation) as the following


[image: Screen Shot 2017-09-04 at 2.22.08 PM.png]
Instead of SA (sum average that was used in pyradiomics), sum entropy (SE)
was used instead. Also, sum variance was defined using sum entropy as well
in this this link
<http://earlglynn.github.io/RNotes/package/EBImage/Haralick-Textural-Features.html>.
Although it makes more sense to use sum average since we're calculating sum
variance, perhaps the source of definition that I mentioned may be
incorrect. so please clarify if you can!


   - Difference variance
      - in pyradiomics documenttaion, difference variance was defined as
      the following

[image: Screen Shot 2017-09-04 at 3.08.11 PM.png]
but in radiomics/glcm.py, line 485-488, difference variance is defined as
diffvar = numpy.sum((pxSuby * ((kValuesDiff[:, None] - diffavg) ** 2)), 0)

which translate to this equation instead..
\textit{difference variance} = \displaystyle\sum^{N_g-1}_{k=0}{(k
-DA)^2p_{x-y}(k)}

Let me know what you think and I am happy to help out by submitting a pull
request if needed. Thanks a lot for your time and help!

Cheers,
Clare Huang


On Mon, Sep 4, 2017 at 1:48 PM Joost van Griethuysen <
notifications@github.com> wrote:

> Remove Sum Variance (GLCM), as this feature is mathematically equal to
> Cluster Tendency. Document the proof of this equality in a new
> documentation section (Excluded Radiomic Features).
>
> As this proof is coded in LaTeX, here is a render of the proof:
> [image: proof_sumvar_clusten]
> <https://user-images.githubusercontent.com/18026947/30038240-66072df0-9190-11e7-8c04-da0289718306.png>
>
> Aside from this equality, there is also a near equality of Joint Average
> and Sum Average (when GLCM is symmetrical, Sum Average = 2 * Joint
> Average). Therefore, disable Sum Average by default in the example
> parameter files and add a warning to the docstring of Sum Average.
>
> Finally, there was a small error in the documentation of the GLCM formulas
> (u_x and u_y erroneously represented as a function of i and j, respectively
> (they are scalar values)).
>
> cc @Radiomics/developers
> <https://github.com/orgs/Radiomics/teams/developers>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/Radiomics/pyradiomics/pull/300
> Commit Summary
>
>    - DOCS: Update GLCM formula documentation
>    - STYL: Update feature function names to better reflect feature names
>    - Remove Sum Variance feature
>
> File Changes
>
>    - *M* docs/features.rst
>    <https://github.com/Radiomics/pyradiomics/pull/300/files#diff-0> (2)
>    - *M* docs/index.rst
>    <https://github.com/Radiomics/pyradiomics/pull/300/files#diff-1> (1)
>    - *A* docs/removedfeatures.rst
>    <https://github.com/Radiomics/pyradiomics/pull/300/files#diff-2> (88)
>    - *M* examples/exampleSettings/Params.yaml
>    <https://github.com/Radiomics/pyradiomics/pull/300/files#diff-3> (33)
>    - *M* examples/exampleSettings/exampleCT.yaml
>    <https://github.com/Radiomics/pyradiomics/pull/300/files#diff-4> (16)
>    - *M* examples/exampleSettings/exampleMR_3mm.yaml
>    <https://github.com/Radiomics/pyradiomics/pull/300/files#diff-5> (56)
>    - *M* examples/exampleSettings/exampleMR_5mm.yaml
>    <https://github.com/Radiomics/pyradiomics/pull/300/files#diff-6> (56)
>    - *M* examples/exampleSettings/exampleMR_NoResampling.yaml
>    <https://github.com/Radiomics/pyradiomics/pull/300/files#diff-7> (56)
>    - *M* radiomics/glcm.py
>    <https://github.com/Radiomics/pyradiomics/pull/300/files#diff-8> (53)
>
> Patch Links:
>
>    - https://github.com/Radiomics/pyradiomics/pull/300.patch
>    - https://github.com/Radiomics/pyradiomics/pull/300.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Radiomics/pyradiomics/pull/300>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGz4vLXYjGtffdEBcdzibewXsukK23Dnks5sfGIigaJpZM4PMPcc>
> .
>

JoostJM commented:
@clarehchao, thanks for pointing out the error in the difference average formula! I already fixed it with a commit in this PR, as I also corrected some other errors in GLCM formula documentation, so another PR is not necessary.

As to your remark on Sum Variance, we are aware of this difference. Indeed, PyRadiomics used to have 2 variants of Sum Variance, but this has been removed as of PR #233, (specifically commit 2042e27). The implementation as defined in the paper by Aerts et al., as well as on the site you mention use the definition as it is in the original Haralick paper (Haralick, R., Shanmugan, K., Dinstein, I; Textural features for image classification; IEEE Transactions on Systems, Man and Cybernetics; 1973(3), p610-621). However, the use of Sum Entropy is most likely the result of a typo (`f8` instead of `f6` in the Sum Variance formula `f7`), as a variance is a sum of the difference between intensity and mean squared.

This choice for Sum Variance using the Sum Average is also consistent with the definition by the Imaging Biomarker Standardisation Initiative (IBSI). Their feature definitions are available [here](https://arxiv.org/abs/1612.07003).

clarehchao commented:
Thank you for the clarification!!! Keep pressing on making awesome package
like pyradiomics!

Clare

On Mon, Sep 4, 2017, 5:25 PM Joost van Griethuysen <notifications@github.com>
wrote:

> @clarehchao <https://github.com/clarehchao>, thanks for pointing out the
> error in the difference average formula! I already fixed it with a commit
> in this PR, as I also corrected some other errors in GLCM formula
> documentation, so another PR is not necessary.
>
> As to your remark on Sum Variance, we are aware of this difference.
> Indeed, PyRadiomics used to have 2 variants of Sum Variance, but this has
> been removed as of PR #233
> <https://github.com/Radiomics/pyradiomics/pull/233>, (specifically commit
> 2042e27
> <https://github.com/Radiomics/pyradiomics/commit/2042e277f7a2ed511d0d4f7f6f73476094d5e073>).
> The implementation as defined in the paper by Aerts et al., as well as on
> the site you mention use the definition as it is in the original Haralick
> paper (Haralick, R., Shanmugan, K., Dinstein, I; Textural features for
> image classification; IEEE Transactions on Systems, Man and Cybernetics;
> 1973(3), p610-621). However, the use of Sum Entropy is most likely the
> result of a typo (f8 instead of f6 in the Sum Variance formula f7), as a
> variance is a sum of the difference between intensity and mean squared.
>
> This choice for Sum Variance using the Sum Average is also consistent with
> the definition by the Imaging Biomarker Standardisation Initiative (IBSI).
> Their feature definitions are available here
> <https://arxiv.org/abs/1612.07003>.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/Radiomics/pyradiomics/pull/300#issuecomment-327042620>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGz4vNuq8dgBjlPozsk1rCXMNNSsIiRjks5sfJTqgaJpZM4PMPcc>
> .
>
